File: src/command/ReceiverStatusUpdate.ts
/*
 * SPDX-License-Identifier: AGPL-3.0-only
 */

export type ReceiverStatus = "open" | "close" | "error" | "connecting";

export type ReceiverStatusUpdateCmd = "receiverStatusUpdate";

export type ReceiverStatusUpdate = {
  cmd: ReceiverStatusUpdateCmd;
  data: {
    status: ReceiverStatus;
  };
};

export function getStatusUpdateMessage(
  status: ReceiverStatus
): ReceiverStatusUpdate {
  return {
    cmd: "receiverStatusUpdate",
    data: {
      status: status,
    },
  };
}

export function getStatusUpdateMessageCmd(): ReceiverStatusUpdateCmd {
  return "receiverStatusUpdate";
}

File: backend/src/command/command_packet/app_command/receiver_status_update.rs
/*
 * Copyright 2021-2023 Golden_Water
 * SPDX-License-Identifier: AGPL-3.0-only
 */

#[derive(serde::Serialize, serde::Deserialize, ts_rs::TS, PartialEq, Eq, Debug, Clone)]
#[serde(rename_all = "camelCase")]
#[ts(
  export,
  export_to = "../frontend/src/share/type/rust/command/commandPacket/appCommand/"
)]
pub struct ReceiverStatusUpdate {
  status: ReceiverStatus,
}

impl ReceiverStatusUpdate {
  pub fn new(status: ReceiverStatus) -> ReceiverStatusUpdate {
    ReceiverStatusUpdate { status }
  }
}

#[derive(serde::Serialize, serde::Deserialize, ts_rs::TS, PartialEq, Eq, Debug, Clone)]
#[serde(rename_all = "camelCase")]
#[ts(
  export,
  export_to = "../frontend/src/share/type/rust/command/commandPacket/appCommand/receiverStatusUpdate/"
)]
pub enum ReceiverStatus {
  Close,
  Connecting,
  Connected,

  Error,
  Reconnecting,

  Interrupted,
}